File: files/jvm/src/main/java/com/swoval/files/ArrayOps.java
package com.swoval.files;

class ArrayOps {
  public interface Filter<T> {
    boolean accept(T t);
  }
  public static <T> boolean contains(final T[] array, final T el) {
    return find(array, el) != null;
  }
  public static <T> T find(final T[] array, final T el) {
    return find(array, new Filter<T>() {
      @Override
      public boolean accept(T t) {
        return el == null ? t == null : t.equals(el);
      }
    });
  }
  public static <T> T find(T[] array, Filter<T> filter) {
    T result = null;
    int i = 0;
    while (result == null && i < array.length) {
      final T el = array[i];
      if (filter.accept(el)) result = el;
      i += 1;
    }
    return result;
  }
}

File: files/js/src/main/scala/com/swoval/files/ArrayOps.scala
package com.swoval.files

object ArrayOps {

  def find[T](array: Seq[T], filter: com.swoval.functional.Filter[T]): T =
    array.find(filter.accept).getOrElse(null.asInstanceOf[T])

  def contains[T](array: Seq[T], el: T): Boolean = {
    var result: Boolean = false
    var i: Int = 0
    while (!result && i < array.length) {
      if (array(i) == el) result = true
      i += 1
    }
    result
  }

}
